activerecord/lib/active_record/associations/has_many_association.rb

```diff
@@ -117,8 +115,25 @@ def delete_or_nullify_all_records(method)
 
         # Deletes the records according to the <tt>:dependent</tt> option.
         def delete_records(records, method)
-          if method == :destroy
+          case method
+          when :destroy
             records.each(&:destroy!)
+            update_counter(-records.length) unless reflection.inverse_updates_counter_cache?
+          when :destroy_async
+            association_class = records.first.class
+            primary_key_column = association_class.primary_key.to_sym
+            ids = records.collect { |assoc| assoc.public_send(primary_key_column) }
```

Comment:
I took this approach from another part of the code. However, I think it may lead to an N+1 query issue. Is it okay to use it like this here?